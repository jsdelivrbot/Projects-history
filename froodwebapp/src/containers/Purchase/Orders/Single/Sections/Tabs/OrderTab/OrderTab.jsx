import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Row, Col } from 'react-flexbox-grid';
import { connect } from 'react-redux';
import { Field, formValueSelector } from 'redux-form';
import {
  FormGridTable,
  Controls,
  ConnectedTopFormModal
} from 'components';
import {
  purchaseOrderSaveRequest,
  purchaseOrderUpdateRequest
} from 'redux-base/actions';
import ConnectedTotalAndNotes from './TotalAndNotes/TotalAndNotes';
import {
  columns,
  prepareOrder,
  prepareOrderForRelease,
  getButtonsByStatus,
} from './orderTabHelpers';
import fields from './modalFields';

const selector = formValueSelector('newPurchaseOrderForm');

const mapStateToProps = state => ({
  loadingPage: state.purchaseOrder.loadingPage,
  vendorId: selector(state, 'vendor.id'),
  initialValues: state.purchaseOrder.initialValues
});

const mapDispatchToProps = {
  purchaseOrderSaveRequest,
  purchaseOrderUpdateRequest
};

const releaseData = {
  email: '',
  subject: '',
  message: ''
};

export class OrderTab extends Component {
  constructor(props) {
    super(props);
    this.state = {
      modalData: releaseData,
      modalVisible: false
    };
  }

  componentWillReceiveProps(nextProps) {
    // to close modal when order released
    if (nextProps.initialValues.id === this.props.initialValues.id &&
      nextProps.status !== this.props.status) {
      this.setState({
        modalVisible: false,
        modalData: undefined
      });
    }
  }

  handleSaveOrder = (formValues, operation) => {
    this.props.purchaseOrderSaveRequest({
      payload: prepareOrder(formValues, operation)
    });
  }

  handleUpdateOrder = (formValues, operation) => {
    if (operation === 'release') {
      this.setState({
        modalVisible: true,
        modalData: {
          ...releaseData,
          ...formValues,
          email: this.props.initialValues.vendor.email
        }
      });
    } else {
      this.props.purchaseOrderUpdateRequest({
        payload: prepareOrder(formValues, operation)
      });
    }
  }

  handleReleaseOrder = (values) => {
    this.props.purchaseOrderUpdateRequest({
      payload: prepareOrderForRelease(values)
    });
  }

  handleToggleModal = () => {
    this.setState({
      modalVisible: !this.state.modalVisible
    });
  }

  render() {
    const {
      loadingPage,
      status,
      vendorId,
      orderInfoDisabled,
      handleSubmit
    } = this.props;

    const {
      modalData,
      modalVisible
    } = this.state;

    const {
      saveButtonVisible,
      saveButtonText,
      saveButtonHandler,
      updateButtonVisible,
      updateButtonText,
      updateButtonHandler,
      cancelButtonVisible,
      cancelButtonText,
      cancelButtonHandler
    } = getButtonsByStatus(
      status,
      this.handleSaveOrder,
      this.handleUpdateOrder
    );

    return (
      <div>
        <ConnectedTopFormModal
          loading={ loadingPage }
          title="Email Purchase Order"
          buttonVisible={ false }
          handleSave={ this.handleReleaseOrder }
          handleToggleModal={ this.handleToggleModal }
          visible={ modalVisible }
          initialValues={ modalData }
          fields={ fields }
        />
        <Row>
          <Col xs>
            <Field
              name="details"
              addNewRowVisible={ !orderInfoDisabled }
              columns={ columns(orderInfoDisabled, vendorId) }
              component={ FormGridTable }
            />
            <ConnectedTotalAndNotes
              disabled={ orderInfoDisabled }
              currency="$"
            />
            <Controls
              saveButtonVisible={ saveButtonVisible }
              saveButtonText={ saveButtonText }
              onSaveButtonClick={ saveButtonHandler && handleSubmit(saveButtonHandler) }
              updateButtonVisible={ updateButtonVisible }
              updateButtonText={ updateButtonText }
              onUpdateButtonClick={ updateButtonHandler && handleSubmit(updateButtonHandler) }
              cancelButtonVisible={ cancelButtonVisible }
              cancelButtonText={ cancelButtonText }
              onCancelButtonClick={ cancelButtonHandler && handleSubmit(cancelButtonHandler) }
            />
          </Col>
        </Row>
      </div>
    );
  }
}

OrderTab.propTypes = {
  // trigger
  loadingPage: PropTypes.bool.isRequired,
  // data
  vendorId: PropTypes.number,
  status: PropTypes.string.isRequired,
  initialValues: PropTypes.object,
  orderInfoDisabled: PropTypes.bool.isRequired,
  // form
  handleSubmit: PropTypes.func.isRequired,
  // redux-base
  purchaseOrderSaveRequest: PropTypes.func.isRequired,
  purchaseOrderUpdateRequest: PropTypes.func.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(OrderTab);
