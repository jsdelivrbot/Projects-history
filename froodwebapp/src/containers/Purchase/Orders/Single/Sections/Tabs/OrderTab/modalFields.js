export default [{
  header: 'To',
  type: 'text',
  name: 'email',
  editableForExistingItem: true,
  editableForNewItem: true,
}, {
  header: 'Subject',
  type: 'text',
  name: 'subject',
  editableForExistingItem: true,
  editableForNewItem: true,
}, {
  header: 'Message to supplier',
  type: 'textarea',
  name: 'message',
  editableForExistingItem: true,
  editableForNewItem: true,
}];
