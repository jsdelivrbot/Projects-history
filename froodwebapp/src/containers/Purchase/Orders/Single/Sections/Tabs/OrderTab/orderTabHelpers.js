import { percentageFormatter, percentageParser } from 'utils';

export const columns = (disabled, vendorId) => [{
  title: 'SKU',
  type: 'skuAutoComplete',
  disabled: disabled || !vendorId,
  vendorId,
  dataIndex: 'sku',
}, {
  title: 'ITEM NAME',
  dataIndex: 'name',
}, {
  title: 'SUPPLIER SKU',
  dataIndex: 'supplierSku',
}, {
  title: 'UOM',
  dataIndex: 'uomName',
}, {
  title: 'AVAILABLE QTY',
  dataIndex: 'availableQty',
}, {
  title: 'PRICE',
  dataIndex: 'price',
  render: (text, record) => record.price && `$${record.price}`
}, {
  title: 'ORDER QTY',
  type: 'number',
  disabled: disabled || !vendorId,
  dataIndex: 'qty',
}, {
  title: 'DISCOUNT',
  type: 'number',
  disabled: disabled || !vendorId,
  formatter: percentageFormatter,
  parser: percentageParser,
  dataIndex: 'discount',
}, {
  title: 'TAX',
  dataIndex: 'tax',
  render: (text, record) => record.tax && `$${record.tax}`
}, {
  title: 'TOTAL',
  dataIndex: 'total',
  render: (text, record) => record.price && (`$${(((record.qty * record.price) - ((record.qty * record.price) * (record.discount / 100))) + record.tax).toFixed(2)}`)
}, {
  type: 'deleteIconColumn',
  allRowsCanBeDeleted: !disabled
}];

export const getButtonsByStatus = (status, saveHandler, updateHandler) => {
  switch (status.toLowerCase()) {
    case 'not saved':
      return {
        saveButtonVisible: true,
        saveButtonText: 'Authorize',
        saveButtonHandler: formData => saveHandler(formData, 'authorize'),
        updateButtonVisible: true,
        updateButtonText: 'Save as Draft',
        updateButtonHandler: formData => saveHandler(formData, 'draft'),
        cancelButtonVisible: false
      };
    case 'draft':
      return {
        saveButtonVisible: true,
        saveButtonText: 'Authorize',
        saveButtonHandler: formData => updateHandler(formData, 'authorize'),
        updateButtonVisible: true,
        updateButtonText: 'Save',
        updateButtonHandler: formData => updateHandler(formData, 'draft'),
        cancelButtonVisible: false
      };
    case 'authorized':
      return {
        saveButtonVisible: true,
        saveButtonText: 'Approve',
        saveButtonHandler: formData => updateHandler(formData, 'approve'),
        updateButtonVisible: false,
        cancelButtonVisible: true,
        cancelButtonText: 'Reject',
        cancelButtonHandler: formData => updateHandler(formData, 'reject')
      };
    case 'approved':
      return {
        saveButtonVisible: true,
        saveButtonText: 'Release',
        saveButtonHandler: formData => updateHandler(formData, 'release'),
        updateButtonVisible: false,
        cancelButtonVisible: false
      };
    case 'released':
    case 'partial received':
      return {
        saveButtonVisible: false,
        updateButtonVisible: false,
        cancelButtonVisible: true,
        cancelButtonText: 'Cancel',
        cancelButtonHandler: formData => updateHandler(formData, 'cancel')
      };
    default:
      return {
        saveButtonVisible: false,
        updateButtonVisible: false,
        cancelButtonVisible: false
      };
  }
};

export const prepareOrder = (formData, operation) => ({
  billToCompany: {
    id: formData.billToCompany.id,
    addressId: formData.billToCompany.address && formData.billToCompany.address.id,
  },
  shipToCompany: {
    id: formData.shipToCompany.id,
    addressId: formData.shipToCompany.address && formData.shipToCompany.address.id,
  },
  vendor: {
    id: formData.vendor.id,
    locationId: formData.vendor.address && formData.vendor.address.id
  },
  id: formData.id && formData.id,
  deliveryDate: formData.deliveryDate,
  operation,
  contactUserId: Number(formData.contactUser.id),
  shippingMethodId: Number(formData.deliveryMethod.id),
  paymentTermsId: Number(formData.paymentTerms.id),
  details: formData.details.map(skuDetail => ({
    sku: skuDetail.sku,
    qty: Number(skuDetail.qty),
    discount: Number(skuDetail.discount),
  })),
  vendorNotes: formData.vendorNotes,
  internalNotes: formData.internalNotes,
  shipping: Number(formData.shipping),
  adjustment: Number(formData.adjustment)
});

export const prepareOrderForRelease = values => ({
  id: values.id,
  email: values.email,
  subject: values.subject,
  message: values.message,
  operation: 'release'
});
