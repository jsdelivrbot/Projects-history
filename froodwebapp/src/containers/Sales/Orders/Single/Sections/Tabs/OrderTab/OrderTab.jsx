import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  FormGridTable,
  FormRadioButtonGroup,
  Controls
} from 'components';
import { connect } from 'react-redux';
import { Row, Col } from 'react-flexbox-grid';
import { Field } from 'redux-form';
import {
  orderSaveRequest,
  orderUpdateRequest
} from 'redux-base/actions';
import ConnectedTotalAndNotes from './TotalAndNotes/TotalAndNotes';
import {
  columns,
  prepareOrder,
  getButtonsByStatus
} from './orderTabHelpers';

const mapStateToProps = state => ({
  status: state.order.initialValues.status,
  initialSkuDetails: state.order.initialValues.skuDetails
});

const mapDispatchToProps = {
  orderSaveRequest,
  orderUpdateRequest
};

const radioButtonValues = [{
  value: false,
  text: 'Single'
}, {
  value: true,
  text: 'Recurring',
}];

export class OrderTab extends Component {
  handleSaveOrder = (formValues, operation) => {
    this.props.orderSaveRequest({
      payload: prepareOrder(formValues, operation)
    });
  }

  handleUpdateOrder = (formValues, operation) => {
    // mark deleted skuDetails
    const skuDetails = [];
    this.props.initialSkuDetails.forEach((skuDetail) => {
      const exist = formValues.skuDetails.some(sd => sd.id === skuDetail.id);
      if (!exist) {
        skuDetails.push({
          ...skuDetail,
          deleted: true
        });
      }
    });

    // delete id for new skuDetails
    formValues.skuDetails.forEach((sd) => {
      skuDetails.push({
        ...sd,
        id: sd.new ? null : sd.id
      });
    });

    this.props.orderUpdateRequest({
      payload: prepareOrder({ ...formValues, skuDetails }, operation)
    });
  }

  render() {
    const {
      // props
      status,
      disabled,
      isNewOrder,
      // redux-form
      handleSubmit,
      onFieldChange
    } = this.props;

    const {
      saveButtonVisible,
      saveButtonText,
      saveButtonHandler,
      updateButtonVisible,
      updateButtonText,
      updateButtonHandler,
      cancelButtonVisible,
      cancelButtonText,
      cancelButtonHandler
    } = getButtonsByStatus(
      status,
      this.handleSaveOrder,
      this.handleUpdateOrder
    );

    return (
      <div>
        <Row>
          <Col xs sm md={ 5 } lg={ 3 }>
            <Field
              name="isRecurring"
              radioButtonValues={ radioButtonValues }
              type="checkbox"
              disabled={ !isNewOrder }
              component={ FormRadioButtonGroup }
            />
          </Col>
        </Row>
        <Row>
          <Col xs sm md lg>
            <Field
              name="skuDetails"
              addNewRowVisible={ !disabled }
              columns={ columns(disabled) }
              component={ FormGridTable }
            />
          </Col>
        </Row>
        <Row>
          <Col xs sm md lg>
            <ConnectedTotalAndNotes
              currency="$"
              disabled={ disabled }
              onFieldChange={ onFieldChange }
            />
          </Col>
        </Row>
        <Controls
          saveButtonVisible={ saveButtonVisible }
          saveButtonText={ saveButtonText }
          onSaveButtonClick={ saveButtonHandler && handleSubmit(saveButtonHandler) }
          updateButtonVisible={ updateButtonVisible }
          updateButtonText={ updateButtonText }
          onUpdateButtonClick={ updateButtonHandler && handleSubmit(updateButtonHandler) }
          cancelButtonVisible={ cancelButtonVisible }
          cancelButtonText={ cancelButtonText }
          onCancelButtonClick={ cancelButtonHandler && handleSubmit(cancelButtonHandler) }
        />
      </div>
    );
  }
}

OrderTab.propTypes = {
  // data
  status: PropTypes.string,
  initialSkuDetails: PropTypes.array,
  // state
  disabled: PropTypes.bool,
  isNewOrder: PropTypes.bool,
  // redux-base
  orderSaveRequest: PropTypes.func.isRequired,
  orderUpdateRequest: PropTypes.func.isRequired,
  // redux-form
  onFieldChange: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(OrderTab);
