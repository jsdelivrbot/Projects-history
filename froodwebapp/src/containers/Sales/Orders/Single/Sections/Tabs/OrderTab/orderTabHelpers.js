import moment from 'moment';
import { percentageFormatter, percentageParser } from 'utils';

export const columns = disabled => [{
  title: 'SKU',
  type: 'skuAutoComplete',
  disabled,
  dataIndex: 'sku',
}, {
  title: 'ITEM NAME',
  dataIndex: 'name',
}, {
  title: 'UOM',
  dataIndex: 'uomName',
}, {
  title: 'AVAILABLE QTY',
  dataIndex: 'availableQty',
}, {
  title: 'PRICE',
  dataIndex: 'price',
  render: (text, record) => record.price && `$${record.price}`
}, {
  title: 'ORDER QTY',
  type: 'number',
  disabled,
  dataIndex: 'qty',
}, {
  title: 'DISCOUNT',
  type: 'number',
  disabled,
  formatter: percentageFormatter,
  parser: percentageParser,
  dataIndex: 'discount',
}, {
  title: 'TAX',
  dataIndex: 'tax',
  render: (text, record) => record.tax && `$${record.tax}`
}, {
  title: 'TOTAL',
  dataIndex: 'total',
  render: (text, record) => record.price && (`$${(((record.qty * record.price) - ((record.qty * record.price) * (record.discount / 100))) + record.tax).toFixed(2)}`)
}, {
  type: 'deleteIconColumn',
  allRowsCanBeDeleted: !disabled
}];

export const getButtonsByStatus = (status, saveHandler, updateHandler) => {
  switch (status.toLowerCase()) {
    case 'not saved':
      return {
        saveButtonVisible: true,
        saveButtonText: 'Confirm',
        saveButtonHandler: formData => saveHandler(formData, 'confirm'),
        updateButtonVisible: true,
        updateButtonText: 'Save as Draft',
        updateButtonHandler: formData => saveHandler(formData, 'draft'),
        cancelButtonVisible: false,
      };
    case 'draft':
      return {
        saveButtonVisible: true,
        saveButtonText: 'Confirm',
        saveButtonHandler: formData => updateHandler(formData, 'confirm'),
        updateButtonVisible: true,
        updateButtonText: 'Save',
        updateButtonHandler: formData => updateHandler(formData, 'draft'),
        cancelButtonVisible: false,
      };
    case 'confirmed':
    case 'backorder':
    case 'partialallocated':
    case 'allocated':
      return {
        saveButtonVisible: true,
        saveButtonText: 'Amend',
        saveButtonHandler: formData => updateHandler(formData, 'amend'),
        updateButtonVisible: false,
        cancelButtonVisible: true,
        cancelButtonText: 'Cancel',
        cancelButtonHandler: formData => updateHandler(formData, 'cancel'),
      };
    case 'shipped':
      return {
        saveButtonVisible: false,
        updateButtonVisible: true,
        updateButtonText: 'Return',
        updateButtonHandler: formData => saveHandler(formData, 'return'),
        cancelButtonVisible: false,
      };
    default:
      return {
        saveButtonVisible: false,
        updateButtonVisible: false,
        cancelButtonVisible: false,
      };
  }
};

export const prepareOrder = (formData, operation) => ({
  id: formData.id,
  addresses: {
    billingAddressId: formData.billingAddress.id,
    shippingAddressId: formData.shippingAddress.id
  },
  channelId: formData.channel.id,
  customerId: formData.customer.id,
  customerNotes: formData.customerNotes,
  deliveryDate: moment(formData.deliveryDate).format('DD-MMMM-YYYY'),
  deliveryMethodId: formData.deliveryMethodId,
  deliverySlotId: formData.deliverySlotId,
  operation,
  paymentTermsId: Number(formData.paymentTerms.id),
  promotionId: formData.promotion && formData.promotion.id,
  skuDetails: formData.skuDetails.map(skuDetail => ({
    id: skuDetail.id,
    sku: skuDetail.sku,
    qty: skuDetail.qty,
    discount: skuDetail.discount,
    deleted: skuDetail.deleted
  }))
});
