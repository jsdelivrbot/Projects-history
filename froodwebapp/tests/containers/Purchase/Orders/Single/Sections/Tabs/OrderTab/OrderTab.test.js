import { OrderTab } from 'containers/Purchase/Orders/Single/Sections/Tabs/OrderTab/OrderTab';
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const { describe, expect, it } = global;

describe('OrderTab', () => {
  const props = {
    vendorId: 1,
    status: 'status',
    readonly: false,
    isNewOrder: false,
    onSave: jest.fn(),
    push: jest.fn(),
    onUpdate: jest.fn()
  };

  it('renders the OrderTab component', () => {
    const component = shallow(<OrderTab { ...props } />);

    expect(component.length).toEqual(1);
  });

  it('capturing Snapshot of OrderTab', () => {
    const component = shallow(<OrderTab { ...props } />);
    const view = toJson(component);
    const actual = expect(view);

    actual.toMatchSnapshot();
  });
});
