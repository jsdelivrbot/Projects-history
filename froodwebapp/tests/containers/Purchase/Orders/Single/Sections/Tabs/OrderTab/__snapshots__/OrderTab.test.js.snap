// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderTab capturing Snapshot of OrderTab 1`] = `
<Row>
  <Col
    xs={true}
  >
    <Field
      addNewRowVisible={true}
      columns={
        Array [
          Object {
            "dataIndex": "sku",
            "disabled": true,
            "title": "SKU",
            "type": "skuAutoComplete",
            "vendorId": 1,
          },
          Object {
            "dataIndex": "name",
            "title": "ITEM NAME",
          },
          Object {
            "dataIndex": "supplierSku",
            "title": "SUPPLIER SKU",
          },
          Object {
            "dataIndex": "uomName",
            "title": "UOM",
          },
          Object {
            "dataIndex": "availableQty",
            "title": "AVAILABLE QTY",
          },
          Object {
            "dataIndex": "price",
            "render": [Function],
            "title": "PRICE",
          },
          Object {
            "dataIndex": "qty",
            "disabled": true,
            "title": "ORDER QTY",
            "type": "number",
          },
          Object {
            "dataIndex": "discount",
            "disabled": true,
            "formatter": [Function],
            "parser": [Function],
            "title": "DISCOUNT",
            "type": "number",
          },
          Object {
            "dataIndex": "tax",
            "render": [Function],
            "title": "TAX",
          },
          Object {
            "dataIndex": "total",
            "render": [Function],
            "title": "TOTAL",
          },
        ]
      }
      component={[Function]}
      deleteColumnVisibleRightSide={false}
      disabled={true}
      name="details"
    />
    <Connect(TotalAndNotes)
      currency="$"
      disabled={true}
    />
    <Controls
      cancelButtonVisible={false}
      onCancelButtonClick={[Function]}
      onSaveButtonClick={[Function]}
      saveButtonText="Save"
      saveButtonVisible={false}
      submitButtonText="Submit"
      submitButtonVisible={false}
    />
  </Col>
</Row>
`;
